pkg/networkservice/common/heal/client.go

```diff
@@ -162,13 +174,22 @@ func (f *healClient) Request(ctx context.Context, request *networkservice.Networ
 	return rv, nil
 }
 
-func (f *healClient) Close(ctx context.Context, conn *networkservice.Connection, opts ...grpc.CallOption) (*empty.Empty, error) {
+func (f *HealClient) Close(ctx context.Context, conn *networkservice.Connection, opts ...grpc.CallOption) (*empty.Empty, error) {
 	rv, err := next.Client(ctx).Close(ctx, conn, opts...)
 	if err != nil {
 		return nil, errors.Wrap(err, "Error calling next")
 	}
 	f.updateExecutor.AsyncExec(func() {
 		delete(f.requestors, conn.GetId())
+		delete(f.closers, conn.GetId())
 	})
 	return rv, nil
 }
+
```

Comment:
Hmm... I'm wondering if there's another way we could address this other than adding a Stop() method (which is going to get lost in the return from New as require us to play crazy type coercion games).  Ideas?

Reply:
I see 2 options here:
-  Since we already have interface-wrapper for [Endpoints](https://github.com/networkservicemesh/sdk/blob/master/pkg/networkservice/chains/endpoint/server.go#L36) probably we could add interface-wrapper for clients with additional method: 
    ```go
    type Disposable interface {
        Dispose() error
    }
    ``` 
    the name is not final, but good one to avoid confusing with `Close`

-  All chain elements should be lazy. So there should be no resource allocation or monitoring in the constructor. We should start all additional goroutines only after the first `Request` called. We can dispose/free/stop all resources when the last `Close` called. 

I like the second option more :)

Reply:
@edwarnicke I implemented the second option, please take a look

Reply:
I will have a look.  I tend to want to limit the API surface, keep it simple :)

Lazy is over all a good thing :)  Will have a look at what you've done :)

Reply:
Fixed with `per-chain` Context